src/main/java/games/strategy/util/Match.java

```diff
@@ -96,70 +89,45 @@ private Match(final Predicate<T> condition) {
 
   /**
    * returns true if all elements in the collection match.
+   * returns false if the collection is empty.
+   */
+  public static <T> boolean allMatchNotEmpty(final Collection<T> collection, final Match<T> match) {
+    return !collection.isEmpty() && allMatch(collection, match);
+  }
+
+  /**
+   * returns true if all elements in the collection match.
    */
   public static <T> boolean allMatch(final Collection<T> collection, final Match<T> match) {
-    // XXX: This method should probably return true when the collection is empty to be consistent with Stream#allMatch
-    // (i.e. the match is vacuously satisfied)
-    if (collection.isEmpty()) {
-      return false;
-    }
-    for (final T current : collection) {
-      if (!match.match(current)) {
-        return false;
-      }
-    }
-    return true;
+    return collection.stream().allMatch(match::match);
   }
 
   /**
    * Returns true if any matches could be found.
    */
   public static <T> boolean anyMatch(final Collection<T> collection, final Match<T> match) {
-    if (collection.isEmpty()) {
-      return false;
-    }
-    for (final T current : collection) {
-      if (match.match(current)) {
-        return true;
-      }
-    }
-    return false;
+    return collection.stream().anyMatch(match::match);
   }
 
   /**
    * Returns true if no matches could be found.
    */
   public static <T> boolean noneMatch(final Collection<T> collection, final Match<T> match) {
-    return !anyMatch(collection, match);
+    return collection.stream().noneMatch(match::match);
   }
 
   /**
    * Returns the number of matches found.
    */
   public static <T> int countMatches(final Collection<T> collection, final Match<T> match) {
-    int count = 0;
-    for (final T current : collection) {
-      if (match.match(current)) {
-        count++;
-      }
-    }
-    return count;
+    return (int) collection.stream().filter(match::match).count();
   }
 
   /**
    * return the keys where the value keyed by the key matches valueMatch.
    */
   public static <K, V> Set<K> getKeysWhereValueMatch(final Map<K, V> map, final Match<V> valueMatch) {
-    final Set<K> rVal = new HashSet<>();
-    final Iterator<K> keys = map.keySet().iterator();
-    while (keys.hasNext()) {
-      final K key = keys.next();
-      final V value = map.get(key);
-      if (valueMatch.match(value)) {
-        rVal.add(key);
-      }
-    }
-    return rVal;
+    return map.keySet().stream().filter(key -> valueMatch.match(map.get(key))).collect(Collectors.toSet());
```

Comment:
NBD, but you can avoid the lookup cost for each element by iterating over the entries instead of the keys:

```java
public static <K, V> Set<K> getKeysWhereValueMatch(final Map<K, V> map, final Match<V> valueMatch) {
  return map.entrySet().stream()
      .filter(entry -> valueMatch.match(entry.getValue()))
      .map(Map.Entry::getKey)
      .collect(Collectors.toSet());
}
```